File: DSA/Arrays/carry-forward/closest-min-max/python/closestMinMax.py
def solve(self, A):
        #find max and min
        #find smallest array starting with min or max and ending with min or max
        min = A[0]
        max = A[0]
        for i in range(1, len(A)):
            if A[i] < min:
                min  = A[i]
            if A[i] > max:
                max = A[i]
        
        length = len(A)
        minIndex = -1
        maxIndex = -1
        for i in range (0, len(A)):
            if A[i] == min:
                minIndex = i
                if maxIndex >= 0:
                    subArrLen = (minIndex - maxIndex + 1)
                    if length > subArrLen:
                        length = subArrLen

            if A[i] == max:
                maxIndex = i
                if minIndex >= 0:
                    subArrLen = (maxIndex - minIndex  + 1)
                    if length > subArrLen:
                        length = subArrLen
        
        return length

File: DSA/Arrays/carry-forward/closest-min-max/javascript/closestMinMax.js
export const closestMinMax = (A) => {
    let min = A[0];
    let max = A[0];

    const N = A.length;

    for (let itr = 1; itr < N; itr++) {
        if (A[itr] > max) {
            max = A[itr];
        }
        if (A[itr] < min) {
            min = A[itr];
        }
    }

    let minSubLen = N;
    let maxIndex = -1;
    let minIndex = -1;

    for (let i = 0; i < N; i++) {
        if (A[i] === min) {
            minIndex = i;
            if (maxIndex >= 0) {
                subLen = minIndex - maxIndex + 1;
                if (subLen < minSubLen) {
                    minSubLen = subLen;
                }
            }
        }
        if (A[i] === max) {
            maxIndex = i;
            if (minIndex >= 0) {
                subLen = maxIndex - minIndex + 1;
                if (subLen < minSubLen) {
                    minSubLen = subLen;
                }
            }
        }
    }
    return minSubLen;
};
